Add state callback to ota component

mmakaay opened the pull request:
# What does this implement/fix? 

This is a re-implementation in relation to PR #1714
In that PR, I created some `on_*`  triggers for the OTA component.
In the review, I was requested to provide a single callback instead.

The callback allows hooking into the OTA upgrade process. In my case, I use it to provide some visual feedback to the user about the OTA upgrade process on the device itself.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (this will require users to update their yaml configuration files to keep working)

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1263
  
# Test Environment

- [x] ESP32
- [ ] ESP8266
- [ ] Windows
- [ ] Mac OS
- [x] Linux

# Explain your changes

This is now a minimal change to provide OTA trigger options.

For testing the callback, I modified the original code a bit and wrapped it in a separate component in my project:
https://github.com/mmakaay/esphome-xiaomi_bslamp2/tree/compatiblity-new-coroutine/components/ota_triggers

I included the following bit of config in my project YAML:
https://github.com/mmakaay/esphome-xiaomi_bslamp2/blob/915643a87d2f9e021f6b3a65079278f3652daaad/example.yaml#L203

This code chains works perfect for me.

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

Commit:
Implemented OTA state callback.

glmnet commented:
Why not adding the triggers to OTA itself? Should be zero impact if user does not use them as linker will not include the code

mmakaay commented:
Because @jesserockz asked me in the other PR to create a single callback instead of the setup that I made.
But maybe I misunderstood the comment and the only issue was that I used four callbacks for driving the four triggers that I made. 
I can add the four triggers (now using a single callback instead of four) back in the code if @jesserockz agrees to that too.

jesserockz commented:
> Because @jesserockz asked me 

I don't recall that sorry. On my mobile, can you link to my comment? That pr has a lot of changes so hard to find reviews

mmakaay commented:
That previous PR has a lot of changes, because I made a booboo somewhere in rebasing or so. Therefore I closed it and started this fresh one.

But I searched for the coment, and I misremembered, it wasn't @jesserockz  but @glm himself who commented. See [this commit](https://github.com/esphome/esphome/pull/1714/commits/24a6ed1afaadd58ba48298f5ee08c586b1781bbb#diff-fdb99e99b3c885a1f37182bc21f504915f117d0b429bd37122dcd81f39d2e4c5R95). The discussion is right at the bottom, saying:

@glmnet: "Please make just one callback like connection state and add an enum to indicate the state as an argument. This would save a bit of memory on the OTA and most likely not many people will benefit from all these.

Take a look at how binary sensor only has the callback manager for state changes and then other callbacks are worked out from there."

So I interpreted that as "bring it down to only a single callback", which is what I did in this commit.
But if the automation triggers (now based on only one callback) are a welcome addition, then I'll add those.

Commit:
Make the linter happy.

jesserockz commented:
Ah ok. 
So i think then the comment on this PR means add it to the OTA config as a trigger so users can run automations via YAML
```
on_ota_state:
  then:
    - ...
```

Commit:
Add triggers for the OTA state changes.

mmakaay commented:
I added the triggers to the ota component code. There are actually four of them. Mainly because the progress and the error state take an extra argument (respectively the progress percentage and the error code).

* on_begin: when the OTA starts
* on_progress: during OTA upgrading, variable `x` is the upgrade progress percentage (0.00 - 1.00)
* on_end: when the OTA completes successfully
* on_error: when the OTA ends in an error, variable `x` contains the error code

To minimize the growth of the firmware, I added a define `USE_OTA_STATUS_CALLBACK` to the code. When enabled, it includes the status callback code in the ota component. So, when it is not enabled, this feature does not get included at all and has no impact on the firmware size.

The code generation will automatically set `USE_OTA_STATUS_CALLBACK` when any of the `on_*` triggers is used.

If this setup is okay, then I will get crackin on the documentation part.

Commit:
Make the linter happy.

glmnet commented:
I might be asking something weird or I might not be making myself clear enough:

What I asked to do was to combine the four `CallbackManager<void(... ` into a single one which was done ok, but that does not mean you have to remove all the four `on_connect`, `on_disconnect`... etc on the yaml side, thats why I pointed how the `binary_sensor`  works with all their triggers `on_press`, `on_release` etc, they all work with a single `CallbackManager` yet they are all defined in binary sensor.

I though the same pattern would apply here, then if you never compile code which uses, e.g. `on_press` then the whole class which handles that trigger should not be included by the linker itself resulting in optimized binary. (The linker or any cpp optimization would never remove the callback managers as they are class members)

So current implementation is a plus on all that with the `#idfef` I'm okay with it, the advantage is already explained, the only reason we tend to not use them too much is because all files are recompiled when the `#define` changes.

glmnet commented:
> Ah ok.
> So i think then the comment on this PR means add it to the OTA config as a trigger so users can run automations via YAML
> 
> ```
> on_ota_state:
>   then:
>     - ...
> ```

adding a "general" `on_state` is tempting as it's almost free. But downsides are that you can introduce a contract here that might lead to a breaking change later, since all current 4 options are already provided it might not be necessary.
See e.g. `on_state` in binary sensor is not triggering for initial states, adding initial states will be a breaking change and this trigger would be better named `on_state_change` given the current behavior

glmnet commented on esphome/components/ota/automation.h:
```diff
@@ -0,0 +1,58 @@
+}  // namespace ota
+}  // namespace esphome
+
+#endif // USE_OTA_STATE_CALLBACK
```
```suggestion
#endif  // USE_OTA_STATE_CALLBACK
```
lint

glmnet committed:
lint

OttoWinter commented on esphome/components/ota/__init__.py:
```diff
@@ -49,3 +85,23 @@ async def to_code(config):
+        await automation.build_automation(trigger, [(int, "x")], conf)
+        use_state_callback = True
+    if use_state_callback:
+        cg.add_define("USE_OTA_STATE_CALLBACK")
```
It's fine as-is, but personally I'd be fine too without the define - I don't think code size is a huge issue here (in esphome generally we have enough of that, plus this would add very little). Memory a bit more, but CallbackManager is only ~12 bytes (the vector) and no issue considering there's only ever one instance of this class.

Just an FYI, no changes needed :)

mmakaay replied:
Okay, thanks for the clarification. My first implementation used 4 CallbackManager for the 4 different triggers, and that got a remark about tuning it down. I went all the way with that define :)

OttoWinter commented on esphome/components/ota/automation.h:
```diff
@@ -0,0 +1,58 @@
+#pragma once
+#ifdef USE_OTA_STATE_CALLBACK
```
This define (note: not a build arg) is defined in esphome/core/defines.h, so before this line there should be an `#include "esphome/core/defines.h"` (it currently works because all usages happen to include defines.h before this file is included)

mmakaay replied:
Lucky me ;-) I changed the code to not depend on luck here.

OttoWinter reviewed (commented):
LGTM, one thing about how defines are handled

Commit:
Added forgotten include of defines.h (thanks @OttoWinter)

mmakaay commented:
@glmnet: "I might be asking something weird or I might not be making myself clear enough"

Don't worry, that was all on me ;-) It's perfectly clear now how I should have interpreted the remarks. The better one knows a system, the easier it is to go along in its philosophies, and had I read your comment now, it would have made perfect sense.

Things I'll change based on the earlier comments:

* I'll drop the `#define`, since it's simply not in line with the rest of the code. It doesn't eat much memory and the recompile "issue" might be annoying users, so let's get rid of that.

* I thought about `on_state` and came to the same conclusion that it wouldn't fully fit. When it would be a state, then `Idle` would need to be added, but that wouldn't really work well with `Error` and `Completed` (since after those, I'd would have to push the state artificially to `Idle`, leading to a choice how long to wait before doing that). So long story short: it wan't a fit for a state, because of its event-y nature. But `on_state_change` does fit its purpose.

**Question for an on_state_change implementation:**

The `IN_PROGRESS` state publishes a float that represents the progress.
The `ERROR` state publishes an integer error code.

Publishing these along with an `on_state_change` trigger gets dirty, since it requires two variables of a different type, which are not used by the other states at all.

Would you agree that `on_state_change` doesn't receive any values, but only is used to handle the bare state change? When the user needs the percentage, then the `on_progress` trigger can be used.

glmnet commented:
I agree on all you say. 

Feel free to leave the #define it's actually good. Just sharing what I believe are common considerations. 

The state change handler can be useful too!

Thanks

Commit:
Accept changes as proposed by 'black'.

mmakaay commented:
Okay, I left the `#define` as-is in the code, and added the `on_state_change` callback. It gets a `state` variable, which contains a string that represents the new state. Tested it with:

```yaml
ota:
  on_state_change:
    then:
      logger.log:
        format: "OTA state change: %s"
        args: [ "state.c_str()" ]
```

If this concludes the code changes, I will continue with the doc updates.

Commit:
Fixed wrongly implemented default for switch.

jesserockz commented:
@mmakaay Can you update the docs and link the PR in the description of this PR?

glmnet commented:
I really don't buy the state being a string, I mean, other parts of code enum values are part of the public api like `COVER_CLOSED = 0.0` etc. 

Not really a big deal, opinionated mostly.

jesserockz commented:
Probably could just forward the enum to the trigger and let the user decide what to do with that...

mmakaay commented:
I wonder what the YAML would look like when I'd forward the enum value instead of the string.
I used a string to make the YAML code read cleanly. It also allows logging the state directly, but for that I can provide a translation function.

I'll look into it and I'll provide the documentation.

glmnet commented:
> I wonder what the YAML would look like when I'd forward the enum value instead of the string.
> 
> I used a string to make the YAML code read cleanly. It also allows logging the state directly, but for that I can provide a translation function.
> 
> 
> 
> I'll look into it and I'll provide the documentation.

Well it will look nicer so long it fits like a glove. 

Otherwise it would look uglier, mostly because you can't switch on a string.

jodur commented:
Is this PR still in progress or worked at??

mmakaay commented:
Yes for sure! I am in the moment a little bit extracted by some flooding that I had due to some excessive rain. During working hours I'm at work, the other time I'm fixing stuff that got cr*pped up due to the flooding. Getting close to finalizing that, so I'm probably going to continue on this PR within a few days.

I know I want it for super cute flashing feedback :-D

mmakaay committed:
Merge branch 'esphome:dev' into add-ota-automation-callback

Commit:
Plain enum for state and use it for on_state.

mmakaay commented:
I switched from an "enum class" to a plain "enum" for the `OTAState` to match the regular way of handling enums. 

The `on_state` trigger now receives the enum instead of a string representation of the state.
To test this, I used the following bit of configuration:

```yaml
ota:
  on_state_change:
    then:
      - if:
          condition:
            lambda: return state == ota::OTA_STARTED;
          then:
            logger.log:
              format: "Here we go, another OTA update!"
      - if:
          condition:
            lambda: return state == ota::OTA_IN_PROGRESS;
          then:
            logger.log:
              format: "Still going strong with the OTA update..."
      - if:
          condition:
            lambda: return state == ota::OTA_COMPLETED;
          then:
            logger.log:
              format: "What a ride, we made it! OTA update completed."
      - if:
          condition:
            lambda: return state == ota::OTA_ERROR;
          then:
            logger.log:
              format: "Oh no, the OTA update crashed!"
```

There are some lint errors. I'll fix those and will then start working on the documentation.

Commit:
Added tests for OTA trigger configuration.

mmakaay commented:
Documentation was added in [documentation PR #1816](https://github.com/esphome/esphome/pull/1816).
Tests were added to `tests/test1.yaml`.

I think this should be it.

jesserockz commented on esphome/components/ota/__init__.py:
```diff
@@ -14,8 +15,27 @@
+CONF_ON_PROGRESS = "on_progress"
+CONF_ON_END = "on_end"
+CONF_ON_ERROR = "on_error"
+CONF_ON_STATE_CHANGE_TRIGGER_ID = "on_state_change_trigger_id"
+CONF_ON_BEGIN_TRIGGER_ID = "on_begin_trigger_id"
+CONF_ON_PROGRESS_TRIGGER_ID = "on_progress_trigger_id"
+CONF_ON_END_TRIGGER_ID = "on_end_trigger_id"
+CONF_ON_ERROR_TRIGGER_ID = "on_error_trigger_id"
```
I think I mentioned this earlier in the review, but you don't need to create a key for every trigger...

Just import and use `CONF_TRIGGER_ID` for all of them

mmakaay replied:
I must have missed that comment, sorry for that.
The code has been updated to use `CONF_TRIGGER_ID` all over the place.

Commit:
Use CONF_TRIGGER_ID for al triggers.

jodur commented:
PR still not accepted merged? I am waiting on this desperately

mmakaay commented:
> PR still not accepted merged? I am waiting on this desperately

Desperately even! 😄
Out of sheer curiousity, may I ask what you are planning to do with these automation options?

jodur commented:
see :[https://github.com/esphome/feature-requests/issues/973](https://github.com/esphome/feature-requests/issues/973)
With this patch i can run a custom lamba function before the upgrade takes place
